### ISSUE TITLE
tf.image.ssim produces nan gradients when utilized as a loss function

### ISSUE THREAD

**Issue by Apprisco** [tensorflow/tensorflow#57353]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.9.1

### Custom Code

No

### OS Platform and Distribution

Ubuntu 18.04

### Python version

3.9.12

### CUDA/cuDNN version

11.3/8.6

### GPU model and memory

Tesla V100

### Current Behaviour?

```shell
tf.image.ssim is intended to be used as an evaluation metric as well as a loss function.
Although normally the code works great, from time to time it produces nan gradients: this was found after ~60+ hours of debugging.
This seems to occur when certain values are too small: even Squared values can cause nan errors (more common with OneDNN optimizations enabled thanks to float errors.)

Here's our implementation of SSIM loss below that solved the nan issue.
```

### Standalone code to reproduce the issue

```shell
Nan inducing code:
 tf.GradientTape(watch_accessed_variables=False) as t:
    t.watch(self.gen_model.trainable_variables)
    gen_loss=tf.image.ssim(real,fake)
g_grads=t.gradient(gen_loss,self.gen_model.trainable_variables)

Our implementation that isn't 100% accurate but stops any nan gradients:
```
    def ssim_loss(self, img1, img2):
        window = self._tf_fspecial_gauss(size=self.WS)  
        (_, _, _, channel) = img1.shape.as_list()
        window = tf.tile(window, [1, 1, channel, 1])
        mu1 = tf.nn.depthwise_conv2d(img1, window, strides = [1, 1, 1, 1], padding = 'VALID')+1e-8
        mu2 = tf.nn.depthwise_conv2d(img2, window, strides = [1, 1, 1, 1], padding = 'VALID')
        mu1_sq = mu1 * mu1+1e-8
        mu2_sq = mu2 * mu2
        mu1_mu2 = mu1 * mu2+1e-8
        img1_2 = img1*img1+1e-8
        sigma1_sq = tf.subtract(tf.nn.depthwise_conv2d(img1_2, window, strides = [1 ,1, 1, 1], padding = 'VALID') , mu1_sq)+1e-8
        img2_2 = img2*img2
        sigma2_sq = tf.subtract(tf.nn.depthwise_conv2d(img2_2, window, strides = [1, 1, 1, 1], padding = 'VALID') ,mu2_sq)
        img12_2 = img1*img2+1e-8
        sigma1_2 = tf.subtract(tf.nn.depthwise_conv2d(img12_2, window, strides = [1, 1, 1, 1], padding = 'VALID') , mu1_mu2)+1e-8
        c1 = (self.k1*self.L)**2+1e-8
        c2 = (self.k2*self.L)**2
        ssim_map = ((2*mu1_mu2 + c1)*(2*sigma1_2 + c2)) / ((mu1_sq + mu2_sq + c1)*(sigma1_sq + sigma2_sq + c2))+1e-8
        return tf.reduce_mean(ssim_map)
```
(credits where credits due, can't remember where we borrowed the impl from.)
```

### Relevant log output

```shell
https://media.discordapp.net/attachments/769691447925669928/1009016713733480469/unknown.png
```
</details>

**Comment by sushreebarsa** [comment]
@Apprisco 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thank you!

**Comment by Apprisco** [comment]
Yes, it happened specifically with us utilizing max_value=1.0. We verified that the images do indeed have min/max values between 0 and 1, so there's no reason for this to happen.

I'm not sure how I'm supposed to report a working code snippet for this: nan losses are already very rare to come by, all I know is that SSIM specifically was causing nan losses. 
https://github.com/jorge-pessoa/pytorch-msssim/issues/2
Similar issues used to happen in pytorch as well.
https://stackoverflow.com/questions/64980914/using-ssim-loss-in-tensorflow-returns-nan-values
another stackoverflow link.

It's a common issue it seems, but very hard to reproduce as it relies on two images being so close together that some gradients return nan (thanks to square gradients :D ). A working code snippet would basically be our entire current project, dataset included, that causes a nan every so often.

**Comment by bhack** [comment]
Check https://github.com/tensorflow/tensorflow/issues/50400#issuecomment-1231502107

**Comment by Apprisco** [comment]
Yes, it seems that default Tensorflow implementation of SSIM as well as MS-SSIM both introduce large issues regarding nan gradients in this manner. Our implementation does include a minor clamping, but consistently reports lower SSIM values than Tensorflow.

**Comment by sachinprasadhs** [comment]
@Apprisco, Seems like it is the duplicate of the above linked issues. Could you please close this issue and track the progress in the above linked issue.

**Comment by Apprisco** [comment]
It's not a duplicate as MS-SSIM and SSIM are different tensorflow functions.

**Comment by bhack** [comment]
They use both the same function, the root cause is the same:

https://github.com/tensorflow/tensorflow/blob/v2.9.1/tensorflow/python/ops/image_ops_impl.py#L4423

https://github.com/tensorflow/tensorflow/blob/v2.9.1/tensorflow/python/ops/image_ops_impl.py#L4541

TF is not offering a loss in the API but just an Op. 
You need to handle/guard this in your loss implementation and eventually contribute the loss in https://github.com/keras-team/keras-cv if you want to share your impl with the community.

**Comment by sachinprasadhs** [comment]
@Apprisco , Could you please refer the above comment and let us know if you need any more details. Thanks!

**Comment by AhmadH-Aglan** [comment]
@bhack I am quite new to this. Can you please explain to me what needs to be done to fix the none gradient. My gradient is all none. Thank you
